File: CommandLine/GCD/GCD/ViewController.m
//
//  ViewController.m
//  GCD
//
//  Created by Prashanna Raghavan on 8/6/17.
//  Copyright © 2017 ASU. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self start];
    // Do any additional setup after loading the view, typically from a nib.
}

-(void)start
{
    dispatch_queue_attr_t qos = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_CONCURRENT, QOS_CLASS_UNSPECIFIED, 0);
    dispatch_queue_t queue = dispatch_queue_create("queue",qos);
    
    dispatch_async(queue, ^{
        for (int i=1; i<=10; i++) {
            NSLog(@"Queue : %d",i);
        }
    });
    
    for (int j=1; j<=10; j++) {
        NSLog(@"Normal : %d",j);
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Projects/Essentials/Map/Map/ViewController.swift
//
//  ViewController.swift
//  Map
//
//  Created by Prashanna Raghavan on 8/16/17.
//  Copyright © 2017 ASU. All rights reserved.
//

import UIKit
import MapKit

class ViewController: UIViewController {
    
    @IBOutlet weak var mapView: PRMapView!
    
    var locationManager = CLLocationManager()
    func checkLocationAuthorizationStatus() {
        if CLLocationManager.authorizationStatus() == .authorizedWhenInUse {
            mapView.showsUserLocation = true
        } else {
            locationManager.requestWhenInUseAuthorization()
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        checkLocationAuthorizationStatus()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        let initialLocation = CLLocation(latitude: 33.423538, longitude: -111.939629)
        mapView.centerMap(location: initialLocation)
        
        let place = Places(title: "AMC Centerpoint 11",
                              locationName: "Centerpoint on Mill",
                              discipline: "Theatre",
                              coordinate: CLLocationCoordinate2D(latitude: 33.423063, longitude: -111.941482))
        
        mapView.addAnnotation(place)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

extension ViewController : MKMapViewDelegate
{
    func mapView(_ mapView: MKMapView, annotationView view: MKAnnotationView, calloutAccessoryControlTapped control: UIControl) {
        let location = view.annotation as! Places
        let launchOptions = [MKLaunchOptionsDirectionsModeKey: MKLaunchOptionsDirectionsModeDriving]
        location.mapItem().openInMaps(launchOptions: launchOptions)
    }
    
    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        if let annotation = annotation as? Places {
            let identifier = "pin"
            var view: MKPinAnnotationView
            if let dequeuedView = mapView.dequeueReusableAnnotationView(withIdentifier: identifier)
                as? MKPinAnnotationView { // 2
                dequeuedView.annotation = annotation
                view = dequeuedView
            } else {
                // 3
                view = MKPinAnnotationView(annotation: annotation, reuseIdentifier: identifier)
                view.canShowCallout = true
                view.calloutOffset = CGPoint(x: -5, y: 5)
                view.rightCalloutAccessoryView = UIButton(type: .detailDisclosure) as! UIView
            }
            return view
        }
        return nil
    }
}
